### ISSUE TITLE
Panic when Reauthenticate with multiple go routine

### ISSUE THREAD

**Issue by NBR41** [gophercloud/gophercloud#1244]
Hi,

it seems that there is a mistake introduced in the commit : 

```
commit 3bb4676e61b8eda86b9d50ab2906708b78017115
Author: Nathan Castelein <garbage-collector@users.noreply.github.com>
Date:   Fri Aug 24 04:02:31 2018 +0200

    Set reauth func to nil before making a new request (#1192)
    
    * Set reauth func to nil before making a new request
    
    * Add tests

diff --git a/provider_client.go b/provider_client.go
index 77368f0..95fa11a 100644
--- a/provider_client.go
+++ b/provider_client.go
@@ -295,7 +295,11 @@ func (client *ProviderClient) Request(method, url string, options *RequestOpts)
                                                seeker.Seek(0, 0)
                                        }
                                }
+                               // make a new call to request with a nil reauth func in order to avoid infinite loop
+                               reauthFunc := client.ReauthFunc
+                               client.ReauthFunc = nil
                                resp, err = client.Request(method, url, options)
+                               client.ReauthFunc = reauthFunc
                                if err != nil {
                                        switch err.(type) {
                                        case *ErrUnexpectedResponseCode:
```

I will try to explain the method to reproduce:
If two go routine need to Reauth with the same client when calling Request, the lock of client is too late in : 

```Go
func (client *ProviderClient) Reauthenticate(previousToken string) (err error) {
	if client.ReauthFunc == nil {
		return nil
	}

	if client.mut == nil {
		return client.ReauthFunc()
	}
	client.mut.Lock()
	defer client.mut.Unlock()

	client.reauthmut.Lock()
	client.reauthmut.reauthing = true
	client.reauthmut.Unlock()

	if previousToken == "" || client.TokenID == previousToken {
		err = client.ReauthFunc()
	}

	client.reauthmut.Lock()
	client.reauthmut.reauthing = false
	client.reauthmut.Unlock()
	return
}
``` 
Some times client.ReauthFunc is nil.
If two GoRoutine try to Reauth,
The first GoRoutine lock client, Reauth, release the Lock.
The second GoRoutine call Reauth, pass the test 
```if client.ReauthFunc == nil {```, 
and in the same time a second GoRoutine set client.ReauthFunc at nil in the method Request:
```
				reauthFunc := client.ReauthFunc
				client.ReauthFunc = nil
				resp, err = client.Request(method, url, options)
				client.ReauthFunc = reauthFunc
```

The second GoRoutine panic when it call : 
```err = client.ReauthFunc()``

**Comment by jtopjian** [comment]
Yes, every time one bug is fixed with Reauth, two more appear :)

Are you able to put together a reproduction of this? It's going to be very hard to resolve without one.

**Comment by NBR41** [comment]
Hi, 
To reproduce, a good is with this test code : 

```go 
package gophercloud

import (
	"net/http"
	"net/http/httptest"
	"testing"
	"time"
)

func TestReauth(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusUnauthorized)
	}))
	defer ts.Close()

	p := &ProviderClient{
		ReauthFunc: func() error {
			return nil
		},
	}
	go p.Request("GET", ts.URL+"/foo", &RequestOpts{})
	p.Request("GET", ts.URL+"/bar", &RequestOpts{})
	time.Sleep(1 * time.Second)
}
```
Then run this  : 
```
~/gocode/src/github.com/gophercloud/gophercloud(master) $ go test -race
==================
WARNING: DATA RACE
Read at 0x00c000098940 by goroutine 6:
  github.com/gophercloud/gophercloud.(*ProviderClient).Request()
      /home/fbernard/gocode/src/github.com/gophercloud/gophercloud/provider_client.go:286 +0xb20
  github.com/gophercloud/gophercloud.TestReauth()
      /home/fbernard/gocode/src/github.com/gophercloud/gophercloud/provider_client_test.go:22 +0x2c0
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162

Previous write at 0x00c000098940 by goroutine 8:
  github.com/gophercloud/gophercloud.(*ProviderClient).Request()
      /home/fbernard/gocode/src/github.com/gophercloud/gophercloud/provider_client.go:302 +0xdd1

Goroutine 6 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:878 +0x650
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1119 +0xa8
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1117 +0x4ee
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1034 +0x2ee
  main.main()
      _testmain.go:42 +0x221

Goroutine 8 (finished) created at:
  github.com/gophercloud/gophercloud.TestReauth()
      /home/fbernard/gocode/src/github.com/gophercloud/gophercloud/provider_client_test.go:21 +0x1fe
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162
==================
--- FAIL: TestReauth (1.01s)
    testing.go:771: race detected during execution of test
FAIL
exit status 1
FAIL	github.com/gophercloud/gophercloud	1.029s
```

As you can see, go show that some race conditions could occur.
Locking this securly, should fix this issue.

**Comment by jtopjian** [comment]
This should be fixed in #1327
